### PR TITLE
core: (another) refactor of read path query processing logic

### PR DESCRIPTION
**Pull request by jussisaurio**
# (another) refactor of read path query processing logic

This PR rewrites our select query processing architecture by moving away from the stateful operator-based execution model, back to a more direct bytecode generation approach that, IMO, is easier to follow. A large part of the bytecode emission itself (`program.emit_insn(...)`) is just copy-pasted from the old implementation (after all, it did _work_), but just structured differently.

## Main Changes

1. Removed the `step()` state machine from operators. Previously, each operator had internal state tracking its execution progress, and parent operators would call `.step()` on their children until they needed to do something else. Reading the code and trying to follow the execution was not very easy, and the abstraction was also too general: there was a lot of unnecessary pattern matching and special casing to make query execution fit the model, when honestly the evaluation of a SELECT without any CTEs or subqueries etc can only go a few different ways.

2. Because of the above change, the main codegen function `emit_program()` now contains a series of linear conditional steps instead of kicking off the state machines with `root_operator.step()`. These steps are just things like: "open the cursors", "open the loops", "emit a record into either the main output or a sorter", etc.

3. The `Plan` struct now (again) contains most of the familiar SELECT query components (WHERE clause, GROUP BY, ORDER BY, etc.) rather than having all of them embedded in a tree of operators. The operator tree now ONLY consists of operators that read from a source table in some way -- so it could just be called a join tree, I guess.

4. There's now `plan.result_columns` which is _ALWAYS_ evaluated to get the final results of a SELECT. Previously the operator state machine thing had a hodgepodge of different ways of arriving at the result row.

5. Removed operators:
   - Removed Filter operator (even in the previous version the Filter operator -- which is really the where clause -- had its predicates pushed down to the table loops, and it didn't really ever exist in the bytecode emission phase anymore)
   - Removed Projection operator (`plan.result_columns`)
   - Removed Limit operator (`plan.limit`)
   - Removed Aggregate operator (`plan.group_by` and `plan.aggregates`)
   - Removed Order operator (`plan.order_by`)

6. Added `ast::Expr::Column` to the vendored sqlite3 parser -- column resolution is now done as early as possible. This eliminates repeated string comparisons during execution. I.e. no need for `resolve_ident_table()` etc

7. Simplified expression result caching by removing the complex, and frankly weird, ExpressionResultCache apparatus. The refactored code handles this by tracking which cursor to read columns from at a given time, and copies values from existing registers if the expression is a computation that has already been done in a previous step of the execution. For example in:
```
limbo> select concat(u.first_name, '-LOL'), sum(u.age) from users u group by concat(u.first_name, '-LOL') order by sum(u.age) desc limit 10;
Michael-LOL|11204
David-LOL|8758
Robert-LOL|8109
Jennifer-LOL|7700
John-LOL|7299
Christopher-LOL|6397
James-LOL|5921
Joseph-LOL|5711
Brian-LOL|5059
William-LOL|5047
```
the query execution engine knows that `concat(u.first_name, '-LOL')` is the second column of the `ORDER_BY` sorter without any complex caching.

**HACK:** For deduplicating expressions in ORDER BY and the SELECT body, the code still relies on expression `==` equality to make those decisions which sucks (e.g. `sum(x) != SUM(x)` -- I've marked the parts where this is used with a TODO, we should have a custom expression equality comparison function instead...). This is not a correctness-breaking thing, but still.

## In short

- No more state machines
- The operator tree is now only a "join tree", pretty much
- No weird general purpose `ExpressionResultCache`
- More direct mapping between SQL operations and generated bytecode -- there's really no harm in carrying the "group by" etc concepts in the bytecode generation phase instead of burying them inside Operators
- When a ResultRow is emitted, it is _always_ done by evaluating `plan.result_columns`, instead of the special-casing and hacks that existed previously
- 600+ LOC removed

### CODE REVIEW

File: core/translate/plan.rs [hunk 1]
```diff
@@ -9,20 +9,34 @@ use sqlite3_parser::ast;
 use crate::{
     function::AggFunc,
     schema::{BTreeTable, Index},
-    util::normalize_ident,
     Result,
 };
 
+#[derive(Debug)]
+pub enum ResultSetColumn {
```
Comment (jussisaurio) [inline core/translate/plan.rs:16]: i'll still try to get rid of this enum. we do need to encode whether a result set column contains aggregates (either directly eg `sum(id)` or indirectly `length(sum(id)`), and ideally, which

File: core/translate/emitter.rs [hunk 2]
```diff
@@ -126,1642 +85,1661 @@ pub struct Metadata {
...
-                                count: group_by.len(),
-                            });
+/// Set up the main query execution loop
+/// For example in the case of a nested table scan, this means emitting the RewindAsync instruction
+/// for all tables involved, outermost first.
+fn open_loop(
+    program: &mut ProgramBuilder,
+    source: &mut SourceOperator,
+    referenced_tables: &[BTreeTableReference],
+    m: &mut Metadata,
```
Comment (pereman2) [inline core/translate/emitter.rs:463]:
```suggestion
    metadata: &mut Metadata,
```

**Review by pereman2 — COMMENTED** [review]
I don't have many opinions for this pr a part from being happy because I don't have to parse all the previous abstractions and now parse a simple one.

File: core/translate/emitter.rs (hunk 2 above)
Comment (jussisaurio) [reply core/translate/emitter.rs]: changed m->metadata everywhere in https://github.com/penberg/limbo/pull/416/commits/db462530f1969b500b97dabdf3845f9ef730625a

File: core/translate/plan.rs [hunk 3]
```diff
@@ -9,20 +9,32 @@ use sqlite3_parser::ast;
...
+    // TODO: encode which aggregates (e.g. index bitmask of plan.aggregates) are present in this column
+    pub contains_aggregates: bool,
+}
+
 #[derive(Debug)]
 pub struct Plan {
-    pub root_operator: Operator,
+    pub source: SourceOperator,
+    pub result_columns: Vec<ResultSetColumn>,
+    pub where_clause: Option<Vec<ast::Expr>>,
```
[the remark this answers is not in the document]
Comment (jussisaurio) [reply core/translate/plan.rs:29]: yep, similarly to how sqlite does it (and i think other engines as well)

commented in https://github.com/penberg/limbo/pull/416/commits/da811dc40380a3ece794de17e0acdfa4ca614dce

**Comment by jussisaurio** [comment]
> I don't have many opinions for this pr a part from being happy because I don't have to parse all the previous abstractions and now parse a simple one.

you and me both... 

image
